Implements aliases support in DCA.

esdrubal opened the pull request:
## Description

Adds support for aliases in DCA.

Fixes #4328 where a DCA warning was emmited while using struct method defined in an impl self.

Closes #4283
Closes #4328

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

mohammadfawaz commented:
I did not understand the purpose of `call_path_decl`. What does it contain exactly? And why do we only need it in certain situations? (example: only for enums but not structs)

esdrubal commented:
> I did not understand the purpose of `call_path_decl`. What does it contain exactly? And why do we only need it in certain situations? (example: only for enums but not structs)

`call_path_decl` is needed for the DCA to be able to retrieve a type alias. For instance in the example below we have EnumScrutinee which contains `enum_ref: DeclRefEnum` already resolved to an enum, it is impossible to infer if we used a type alias from that. So I added `call_path_decl` which may be an alias or an enum decl.

https://github.com/FuelLabs/sway/blob/9b80160f681df58138079e841920e987518268b6/sway-core/src/language/ty/expression/scrutinee.rs#L27-L31

Structs should also have it, I probably forgot to add tests cases for structs, I will check that.

esdrubal commented:
Fixed merging conflicts. And added one more test case and a fix for it.

The reason we have `call_path_decl` in `EnumScrutinee` and not in `StructScrutinee` is because we build an `UnsafeDowncast` from `EnumScrutinee`.  The DCA uses `UnsafeDowncast` to connect match statements to the enum variants. In case we use an alias in the enum we need `call_path_decl` to contain it.
